plugins/techdocs-node/src/stages/generate/helpers.ts

```diff
@@ -159,22 +159,43 @@ export const MKDOCS_SCHEMA = DEFAULT_SCHEMA.extend([
  */
 export const getMkdocsYml = async (
   inputDir: string,
+  entity: Entity,
 ): Promise<{ path: string; content: string }> => {
   let mkdocsYmlPath: string;
   let mkdocsYmlFileString: string;
   try {
     mkdocsYmlPath = path.join(inputDir, 'mkdocs.yaml');
-    mkdocsYmlFileString = await fs.readFile(mkdocsYmlPath, 'utf8');
-  } catch {
-    try {
-      mkdocsYmlPath = path.join(inputDir, 'mkdocs.yml');
+    if (await fs.pathExists(mkdocsYmlPath)) {
       mkdocsYmlFileString = await fs.readFile(mkdocsYmlPath, 'utf8');
-    } catch (error) {
-      throw new ForwardedError(
-        'Could not read MkDocs YAML config file mkdocs.yml or mkdocs.yaml for validation',
-        error,
-      );
+      return {
+        path: mkdocsYmlPath,
+        content: mkdocsYmlFileString,
+      };
+    }
+
+    mkdocsYmlPath = path.join(inputDir, 'mkdocs.yml');
+    if (await fs.pathExists(mkdocsYmlPath)) {
+      mkdocsYmlFileString = await fs.readFile(mkdocsYmlPath, 'utf8');
+      return {
+        path: mkdocsYmlPath,
+        content: mkdocsYmlFileString,
+      };
     }
+
+    // No mkdocs file, use default
+    const defaultMkdocsContent =
```

Comment:
What are your thoughts on making this generated based on a handlebars template that say is configured in the `app-config.yaml`? It could pass some well known variables for the entity ref etc in the context the template can use to render a file. This would help this be useful for orgs that are using more then just `mkdocs-techdocs-core`, etc.

Also, I think it also be useful to extract the logic of generating the `mkdocs.yaml` file out so the `techdocs-cli serve` command can also use this. As I recall it does something different but it would be great to be able to preview any repo that has a docs folder without requiring a `mkdocs.yaml` file. The CLI could clean this file up in `SIGTERM` etc.

We a custom wrapper CLI around the `techdocs-cli` does most of this already, just completely isolated from the backstage code base. I've been meaning to upstream it, but you've beat me to it. Would be happy to donate portions to help.

Reply:
For sure, would be happy to work with you on this. I left a comment in the related issue I had logged around this: https://github.com/backstage/backstage/issues/15413#issuecomment-1366048970. Would it make sense to hammer out the architecture for this there?

Reply:
Thinking about this, I think option #2 (`mkdocsConfigTransformer`) from my thoughts in https://github.com/backstage/backstage/issues/15413#issuecomment-1368994736 might be the way forward. This can be extended on top of this PR though.

However, I think we should consider at a minimum cleaning up any newly created `mkdocs.yml` files (as per my comments in #15413). This is likely more of an issue when running `techdocs-cli serve` for previewing the docs sites rather then the CLI build in CI or local Backstage builds as the files are unlikely to be committed to source control in those cases.

I wonder if the `getMkdocsYml` could be modified to return more contextual information. For example it could possibly return if the file existed or not and the `techdocs-cli` could then cleanup the file on `process.exit` if the file is new. Additionally, I think `getMkdocsYml` could be the place to hook in and provide a transformer hook. However, some refactoring to separate data transformation from the physical file might be ideal to split concerns. Again, the transform stuff could be added later.

Reply:
I've added a `configExists` property as part of the output from `getMkdocsYml`. Not super happy with the name so open to suggestions.

Where I'm stuck now is how I can hook into the `process.exit` in a nice way and use this new property to decide if I should delete the `mkdocs.yml` file or not. When doing this in Backstage itself I don't think this is a concern but from the CLI as you said it needs to be done. Here's the places I think I need to do this in:

- `generate` commnand: I would call `getMkdocsYml` somewhere at the start and then probably at the bottom: https://github.com/backstage/backstage/blob/05c010a7dc1d994c36e0c78d9942836d8c784037/packages/techdocs-cli/src/commands/generate/generate.ts#L111 - check if I should do the delete
- `serve` and `serve:mkdocs` commands: both use `waitForSignal` (https://github.com/backstage/backstage/blob/05c010a7dc1d994c36e0c78d9942836d8c784037/packages/techdocs-cli/src/lib/run.ts#L69) which I'm not sure how best to add what I need.

At this point I could really use your advice and guidance @andrewthauer, when you have some time, please

Reply:
I actually just thought of an alternative approach but I'm actually not sure it will work when using docker. However, I'll run it by you just in case maybe there is something to go off.

Basically, instead of removing any new config files created, just generate a `tmp` file for `mkdocs.yml` (e.g. `/tmp/xxxxx.xxx` (probably using nodes os:mkdtemp or something). If you add an absolute `docs_dir` in this temporary config to the `/docs` directory you can then invoke `mkdocs serve/build -f ${tempFile}`. However, when using docker this may not work because the file isn't volume mounted along with the source repo. However, I suppose it could also volume mount the config file perhaps (but haven't tested).

Reply:
As for the the exit approach, yeah all I think you would need to do is return a flag indicating if the file is new or not. I agree `configExists` is a bit misleading because it will exist after the call to `getMkDocsYml`. Maybe something like `configIsNew` or maybe `configIsTemporary` might be more clear? I "think" it might look something like this:

```ts
const { path, configIsNew } = await getMkDocsYml();

if (configIsNew) {
  process.on('exit', async () => {
    fs.rmSync(path, {});
  };
}
```

I believe you'd need to do the same for each CLI command and/or create a shared function that does this.

Reply:
Alright, got back to this. First I've changed the property from the confusing and unclear `configExists` to `configIsTemporary`. Feel like that makes this much more obvious.

I've decided to stick with the `process.on('exit)` to handle the delete. I like the `tmp` file idea I'm just not sure I can get that working. Maybe in another PR I'll give it a try

Reply:
Makes sense. The `tmp` approach is definitely more work and testing. Refactoring this later to use a different approach will unlikely require any changes to the API surface. Although the `configIsTemporary` flag may become obsolete.